### PR TITLE
Enable Content Security Policies

### PR DESCRIPTION
**Pull request by mishaschwartz**
## Motivation and Context

For security reasons, we should implement some CSP policies. This PR makes a first attempt at that.

Ideally we want to enforce strong policies using nonce keys in all cases. However, we currently depend on at least two javascript libraries (`heic2any` and `mathjax`) that require more lax CSP policies. `MathJax` at least has indicated that it is unlikely to improve (https://github.com/mathjax/MathJax/issues/1988). Further discussion on the best way forward is needed.

TODO: further spot-checks of pages to see if the new CSP rules block things that shouldn't be blocked

## Your Changes

**Description**:

**Type of change** (select all that apply):

 

- [x] Other (please specify): security update

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have updated the Changelog.md file. 
- [x] I have described any required documentation changes below. 

### Required documentation changes (if applicable)

https://github.com/MarkUsProject/Markus/wiki/Pending-Changes#pr-5186

### CODE REVIEW

**Comment by david-yz-liu** [comment]
@mishaschwartz I'm testing these changes now, and have encountered 3 issues:

1. The flash message containers don't close when you click on the "X" (it seems there's an `onclick` attribute set there).
2. Under the exam template settings page, when you click on the "Add Template Division" link, 2 blank rows get added instead of just 1. (There's another bug which also occurs on master: when there's more than one exam template, clicking "Add Template Division" adds an empty row to every exam template's division table, not just the one that was clicked. You don't need to fix that here, just something I noticed.)
3. Under the automated test settings page, there's an error when adding a new tester. In the JS console: "Error compiling schema, function code: ..."

I'll try to do more testing on this today, but wanted to get these to you first!

**Comment by mishaschwartz** [comment]
Amazing! Thanks for finding those

File: app/views/annotation_categories/_annotation_list.html.erb [hunk 1]
```diff
@@ -6,10 +6,17 @@
     assignment_id: annotation_category.assignment.id,
     annotation_category_id: annotation_category.id
   ) %>
+  <%= javascript_tag nonce: true do %>
+    $(function() {
+      $('#add-annotation').click(function() {
+        add_annotation_prompt('<%= path %>')
```
Comment (mishaschwartz) [inline app/views/annotation_categories/_annotation_list.html.erb:12]: `e.preventDefault()`

And check to make sure we're doing this for all other cases too

**Review by david-yz-liu — COMMENTED** [review]
Looking good! Left a few inline comments, and found two other issues;

1. Scanned exam issues: I think the CSP is blocking the PDF loading for the "Fix Scans" and "Assign Scans" pages.
2. In the codeviewer, there's a dynamically-generated toolbar with buttons for "view plan" and increasing/decreasing font size. The toolbar shows up but the buttons don't work.

File: app/views/tags/index.html.erb [hunk 2]
```diff
@@ -2,11 +2,16 @@
...
                handlers: [:erb] %>
-  <script>
+  <%= javascript_tag nonce: true do %>
     document.addEventListener('DOMContentLoaded', () => {
       makeTagTable(document.getElementById('tags_table'), {assignment_id: <%= @assignment.id %>});
     });
-  </script>
+    $(function() {
+      $('#add-tag').click(function() {
+        modal_create_new.open()
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply app/views/tags/index.html.erb:11]: @david-yz-liu  So I have a question about these ones.... the original code for this (line 21) did not prevent the default action (ie. no return false and the link_path is #). If we add `preventDefault` here we will be changing the behaviour. 

I don't think that's necessarily a bad thing (opening modals should, in general, not bring you to the top of the page first) but this isn't the behaviour we had before. 

Do you have any opinion on this?

[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/views/tags/index.html.erb:11]: Great point. Yeah I think the previous behaviour wasn't deliberate, just an oversight on my part. I think it's best to make sure opening modals doesn't jump to the top. Making this consistent is a benefit of your work on this PR! :)

**Pull request merged**